README.md
[![Open in Visual Studio Code](https://classroom.github.com/assets/open-in-vscode-c66648af7eb3fe8bc4f294546bfd86ef473780cde1dea487d3c4ff354943c9ae.svg)](https://classroom.github.com/online_ide?assignment_repo_id=7943839&assignment_repo_type=AssignmentRepo)
<div id="top"></div>
<!--
*** Thanks for checking out the Best-README-Template. If you have a suggestion
*** that would make this better, please fork the repo and create a pull request
*** or simply open an issue with the tag "enhancement".
*** Don't forget to give the project a star!
*** Thanks again! Now go create something AMAZING! :D
-->



<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]



<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/github_username/repo_name">
    <img src="images/logo.png" alt="Logo" width="80" height="80">
  </a>

<h3 align="center">myawesometool</h3>

  <p align="center">
    A CLI tool to automatically check for dependency of a specified package and then create a PR updating the package.
    <br />
    <a href="https://github.com/dyte-submissions/dyte-vit-2022-Aritra779"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/dyte-submissions/dyte-vit-2022-Aritra779">View Demo</a>
    ·
    <a href="https://github.com/dyte-submissions/dyte-vit-2022-Aritra779/issues">Report Bug</a>
    ·
    <a href="https://github.com/dyte-submissions/dyte-vit-2022-Aritra779/issues">Request Feature</a>
  </p>
</div>



<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
        <li><a href="#how_to_use">How to use</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>

___

<!-- ABOUT THE PROJECT -->
## About The Project

Automatic Pull requests like below

![Product Name Screen Shot][product-screenshot]

*Currently works with only github.*

This project is about building a simple CLI tool that will look through a file(currently only .csv file) and let the user know about the version of specified dependency of a project/repo listed in the file. It also writes the same info onto another file.

The link to the repository can be any one of the following format:
(taken straight out of [github-url-to-object](https://www.npmjs.com/package/github-url-to-object) page)
<pre>
  * 'github:monkey/business'
  * 'https://github.com/monkey/business'
  * 'https://github.com/monkey/business/tree/master'
  * 'github:monkey/business#nachos'
  * 'https://github.com/monkey/business/tree/master/nested/file.js'
  * 'https://github.com/monkey/business.git'
  * 'http://github.com/monkey/business'
  * 'git://github.com/monkey/business.git'
  * 'git+https://github.com/monkey/business.git'
</pre>

In addition to the above mentioned task if the version is below specified version it'll automatically `fork`, `clone`, `update`, `push` and `create a PR` to the original repo. Provided we tell it to do so.
<p align="right">(<a href="#top">back to top</a>)</p>

### Built With
- Platforms and Registry
  * [nodeJS](https://nodejs.org/)
  * [npm](https://www.npmjs.com/)
- Libraries(for Core Functionality)
  * [commander](https://www.npmjs.com/package/commander)
  * [chalk](https://www.npmjs.com/package/chalk)
  * [dotenv](https://www.npmjs.com/package/dotenv)
  * [octokit/rest](https://www.npmjs.com/package/@octokit/rest)
  * [github-url-to-object](https://www.npmjs.com/package/github-url-to-object)
- Libraries (for Testing)
  * [nock](https://www.npmjs.com/package/nock)
  * [mocha](https://www.npmjs.com/package/mocha)
  * [chai](https://www.npmjs.com/package/chai)

<p align="right">(<a href="#top">back to top</a>)</p>


<!-- GETTING STARTED -->
## Getting Started

To get a local copy up and running follow these simple example steps.

### Prerequisites

In order to use this project/ repo / tool there are certain pre-requisites. 
* node
* npm
  ```sh
  npm install npm@latest -g
  ```
* A code Editor (e.g., Visual Studio, VS Code, Sublime Text, Atom etc.)

### Installation

1. Get your personal access token from github(required only when using the `-u` option. Not required for the `-i` option). Follow this for creating your PAT(Personal Access Token) [Creating a personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
2. Clone the repo
   ```sh
   git clone https://github.com/dyte-submissions/dyte-vit-2022-Aritra779.git
   ```
3. Install NPM packages
   ```sh
   npm install
   ```
4. (Optionally required for `-u` option) Enter your PAT(Personal Access Token) in a `.env` file in `G_TOKEN` variable. Change the name everywhere if you wish.
   ```
   G_TOKEN = 'ENTER YOUR PAT'
   ```

<p align="right">(<a href="#top">back to top</a>)</p>

### How to use

* Move into the project directory from command line / shell / bash
  ```sh
  cd dyte-vit-2022-Aritra779
  ```
* For help run
  ```sh
  myawesometool -h
  ``` 
  or 
  ```sh
  myawesometool --help
  ```
* For reading a CSV file and listing out dependency version satisfiability status run 
  ```sh
  myawesometool -i <somefile.csv> <package_name@version>
  ``` 
*  For creating an automated PR(pull request) along with the above task run
    ```sh
    myawesometool -update -i <somefile.csv> <package_name@version>
    ``` 
    **This requires github PAT(Personal Access Token).**

    *Also note that this will not update the mentioned package if the mentioned version isn't in the compatibility range mentioned in the package.json*
* For forcefully updating a package even if it might break certain features run
  ```sh
  myawesometool -i -u -f <somefile.csv> <package_name@version>
  ```
* For starting at a specified row index run
  ```sh
  myawesometool -i -s4 <somefile.csv> <package_name@version>
  ```
  OR
  ```sh
  myawesometool -i --start=4 <somefile.csv> <package_name@version>
  ```
  If the csv file has no header then pass `--start=-1`
* If you own all the repositories mentioned in the csv file or if you have write access or branch creation access to them then run
  ```sh
  myawesometool -i -u -o <somefile.csv> <package_name@version>
  ```
  Use this option when forking isn't necessary.

  Currently mixture of self owned and not owned repository cannot be process together. Best action plan in that case would be to keep them in two differnt files and then use the tool.
* ```sh
  npm test
  ``` 
  currently has only 4 tests. Fork test and Pull test test the correspoding API calls. The API call is mocked. You can test it out without a worry. The other two tests are for testing out API call to get the package.json file of any repo. These two calls are also mocked. Hence you can test it out without a single worry.
* Don't run it outside of the project directory for the time being (*will be looked into later*)
* This tool expects the first column of the csv file to be the `name` of the repository, the second column to be the `link of the repository`. It'll also add the `version` in the 3rd column, `version satisfiability condition` in 4th column and the `generated PR link` in the 5th column. 
* **Note:** *After the task has been performed the auto generated directory named `Temporary_Directory` will be automatically deleted. But this takes some time, huge lot of time in some cases. Also it doesn't report the deletion progress. It'll only report when there's an error or when the complete deletion task is done. If you don't want to wait around for the deletion task then it's best to stop the process there and manually delete the directory. No harm will be done to the overall task since the core task has already been completed by that time.* 
* There's a sample_d.csv file with two repo links. First repo (named `test`) have `axios 0.25.0` and `react 18.0.0`. Second one doesn't have `react` but have `axios 0.23.0`. So for testing PR generation test with anything higher than the specified versions.
<!-- USAGE EXAMPLES -->

## Usage

<div align = 'center'>

![i option Screen Shot with lower version][i_option_ss1]

![i option Screen Shot with higher version][i_option_ss2]

![npm test Screen Shot][npm_test_ss]

![-u Option Screen Shot1][u_option_ss1]

![-u Option Screen Shot2][u_option_ss2]

![-u Option Screen Shot3][u_option_ss3]

</div>

<p align="right">(<a href="#top">back to top</a>)</p>


<!-- ROADMAP -->
## Roadmap
- [x] ~~Basic Documention (High Priority)~~
- [ ] More Documention (When More Features Arrive)
- [ ] `npm test` (Normal Priority)
- [x] ~~Move from *'have to work anyhow'* to *'should be logically sound'* (High Priority)~~
- [ ] Reduce dependency (Low Priority)
    - [x] ~~csv-parser (Low Priority)~~
    - [x] ~~csv-writer (Low Priority)~~
    - [x] ~~get-repo-package-json (High Priority)~~
    - [ ] github-url-to-object (Not Likely)
    - [ ] chalk (probably Not)
    - [ ] commander (When there's a better option available)
- [ ] Robustness (Normal Priority)
- [ ] More features I can't think of right now (Low Priority)

<p align="right">(<a href="#top">back to top</a>)</p>

## Changelog
### v1.2.0

- Added an additional flag `-s | --start` for specifying which row (0 indexed) of the file to start execution from
- Added an additional flag `-o | --own` for specifying whether the user owns all the repository listed in the csv file. If this flag is passed forking won't be done.
- Added a request limit checking logic that'll automatically stop the execution if the request limit per hour (according to github API) is about to be reached.
- Added a condition that'll check if the github PAT is defined in `.env` file under `G_TOKEN` variable. If not then using `-u | --update` will not perform the said task.

See the full list of changelog listed on [*CHANGELOG.md*](./CHANGELOG.md)
<p align="right">(<a href="#top">back to top</a>)</p>

<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- CONTACT -->
## Contact

Aritra Dutta - aritra.dutta.366@gmail.com

Project Link: [https://github.com/dyte-submissions/dyte-vit-2022-Aritra779](https://github.com/dyte-submissions/dyte-vit-2022-Aritra779)

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- ACKNOWLEDGMENTS -->
## Acknowledgments

* []()
* []()
* []()

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/github_username/repo_name.svg?style=for-the-badge
[contributors-url]: https://github.com/dyte-submissions/dyte-vit-2022-Aritra779/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/github_username/repo_name.svg?style=for-the-badge
[forks-url]: https://github.com/dyte-submissions/dyte-vit-2022-Aritra779/network/members
[stars-shield]: https://img.shields.io/github/stars/github_username/repo_name.svg?style=for-the-badge
[stars-url]: https://github.com/dyte-submissions/dyte-vit-2022-Aritra779/stargazers
[issues-shield]: https://img.shields.io/github/issues/github_username/repo_name.svg?style=for-the-badge
[issues-url]: https://github.com/dyte-submissions/dyte-vit-2022-Aritra779/issues
[license-shield]: https://img.shields.io/github/license/github_username/repo_name.svg?style=for-the-badge
[license-url]: https://github.com/dyte-submissions/dyte-vit-2022-Aritra779/blob/master/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://www.linkedin.com/in/aritra-dutta-6943a1222/
[product-screenshot]: images/pull_proof.png
[i_option_ss1]: images/i_option_1.png
[i_option_ss2]: images/i_option_2.png
[npm_test_ss]: images/npm_test.png
[u_option_ss1]: images/u_flag_ss1.png
[u_option_ss2]: images/u_flag_ss2.png
[u_option_ss3]: images/u_flag_ss3.png


CHANGELOG.md
# Changelog

## v1.2.0

### Added
- Added an additional flag `-s | --start` for specifying which row (0 indexed) of the file to start execution from
- Added an additional flag `-o | --own` for specifying whether the user owns all the repository listed in the csv file. If this flag is passed forking won't be done.
- Added a request limit checking logic that'll automatically stop the execution if the request limit per hour (according to github API) is about to be reached.
- Added a condition that'll check if the github PAT is defined in `.env` file under `G_TOKEN` variable. If not then using `-u | --update` will not perform the said task.


### Changed
- Fixed typos
- Changed the core logic. Previously all the content would have been read first and then processing would start. Now the processing happends line by line i.e., it reads a line, takes the necessary action and then reads the next line. For hugely large files (doubltful anyone would have large file) it'll increase performance.
- Now the output won't be written on the same file. Instead a new file will be created with ***_output*** suffix. This is done to make sure even if someone executes the command by mistake they'll still have access to the previous data in the original file. If the file already exists then it's not overridden instead new content is appended. 

### Removed
- [csv-parser](https://www.npmjs.com/package/csv-parser)
- [csv-writer](https://www.npmjs.com/package/csv-writer)
- Removed the line that would report current package index out of all packages in the form *`[current_index / total_count]`*. Since the file content isn't pre-read anymore this won't be possible.

## v1.1.1

### Changed
- Fixed typos
- Fixed the images in readme not showing up (hopefully this time it's fixed)
- Fixed the output csv file not having the correct link to the generated pull request

## v1.1.0

### Added
- Added more "Built With" frameworks/libraries in readme
- Added an additional flag `-f | --force` for force updating the package even if it might break some features
- Added a test for testing `pull` under `Fork-> Clone -> Push -> Pull` block
- Added two tests under `Getting the package.json of a Repository` for testing the said action
- More robust version checking. Supports packages with semantic versioning with formats such as
    - ^1.5.6
    - ~2.4.5
    - 1
    - 2.x
    - 17.0
    - 17.0.x
    - 1.6.5 - 2.5.6
- After all entries have been checked and necesary actions have been done the `Temorary_Directory` will be deleted.


### Changed
- Fixed typos
- Fixed the images in readme not showing up
- Moved from *'have to work anyhow'* to *'should be logically sound'* (Hopefully)
- Changed the name of a test block from `GITHUB REST API` to `Fork-> Clone -> Push -> Pull`  (though there are no tests for cloning and pushing)
- `Temporary_Directory` won't be pushed to github anymore. (Even if it's not automatically deleted)
- Changed the priotity of `npm test` from *High* to *Normal*
- Changed the priority of *reduce dependency* from *Noraml* to *Low*
### Removed
- [get-repo-package-json](https://www.npmjs.com/package/get-repo-package-json)
- test for get-repo-package-json

## v1.0.0

### Added
- Added this changelog :)
- Back to top links
- Added "Built With" frameworks
- Changed table of contents to start collapsed
- Added checkboxes for major features on roadmap
- Added the entire project except for the pre-build template

test/index.js
import { expect} from 'chai';
import nock from 'nock';
import getResponse from './testData/getResponse.json' assert { type: 'json' };
import getResponse2 from './testData/getResponse2.json' assert { type: 'json' };
import forkResponse from './testData/forkResponse.json' assert {type : 'json'};
import pullPesponse from './testData/pullResponse.json' assert {type : 'json'};
import {fork, clone_and_push, pull} from '../commands/updateDependency.js';
import dotenv from 'dotenv';
import { Octokit } from '@octokit/rest';

dotenv.config();
nock.disableNetConnect();

const getApi = (path) => {
  return nock('https://api.github.com').get(path);
}

const postApi = (path) => {
  return nock('https://api.github.com')
    .intercept(path, 'POST');
}



describe('Getting the package.json of a Repository', () => {
  const octokit = new Octokit({});
    it("testing without ref argument", async () =>{
      let mock = getApi(/\/repos\/.+\/.+\/contents\/package.json(\?ref=.+)?/i).reply(200, getResponse);
      let response = await octokit.request(`GET /repos/someone/somerepo/contents/package.json`, {
        owner: 'someone',
        repo: 'somerepo',
        path: './package.json',
      })
      expect(response).to.be.an('object');
      expect(response.status).to.be.equal(200);
      expect(response.data).to.be.an('object');
      expect(response.data.name).to.be.equal('package.json');
      expect(response.data.content).to.be.a('string');
    })

    it("testing with ref argument", async () =>{
      let mock = getApi(/\/repos\/.+\/.+\/contents\/package.json(\?ref=.+)?/i).reply(200, getResponse2);
      let response = await octokit.request(`GET /repos/someone/somerepo/contents/package.json`, {
        owner: 'someone',
        repo: 'somerepo',
        path: './package.json',
        ref : 'updates'
      })
      expect(response).to.be.an('object');
      expect(response.status).to.be.equal(200);
      expect(response.data).to.be.an('object');
      expect(response.data.name).to.be.equal('package.json');
      expect(response.data.content).to.be.a('string');
    })
})

describe('Fork-> Clone -> Push -> Pull' , () => {
  it('fork test', async () => {
    try{
      let mock = postApi(/\/repos\/.+\/.+\/forks/i).reply(202, forkResponse);
    let response = await fork('someone','Hello-World');
    expect(response.status).to.equal(202);
    expect(response.data['full_name']).to.be.a('string');
    expect(response.data['html_url']).to.be.a('string');
    expect(response.data.parent).to.be.an('object');
    expect(response.data.source).to.be.an('object');
    }catch(err){
      console.log(err)
    }
  })

  it('pull test' , async () => {
    let mock = postApi(/\/repos\/.+\/.+\/pulls/i).reply(201, pullPesponse);
    try{
      let response = await pull('main-repo-owner','Main_Repo', 'forked-repo-owner', 'axios@0.25.0');
      expect(response.status).to.be.equal(201);
      expect(response.data).to.be.an('object');
      expect(response.data.url).to.be.a('string');
    }catch(err){
      console.log(err)
    }
    
  })
})

index.js
#! /usr/bin/env node
import { program } from 'commander';
import checkDependency from './commands/checkDependency.js';

program
    .argument('<File_Name>','name of the csv file(along with extension)')
    .argument('<Package_Name@Version>' , 'package / library to check along with version to check against in the format <package_Name@package_version>')
    .version('1.2.0')
    .requiredOption('-i, --input', 'takes input in the form of a .csv file and a package/library name along with version')
    .option('-u, --update' , '(optionally fork) -> clone -> update -> push -> create PR')
    .option('-f, --force', 'forcefully runs the task even if it might involve some breaking changes')
    .option('-s --start <Row_Index>',"Mention which row(0 indexed) to start from. If file has no header mention -1. Please use this flag as the last flag.",0)
    .option('-o --own', 'You\'re the owner(or have write access) of all the repositories listed in the file. Hence no forking will be done.')
    .action((first, second, options) => {
        checkDependency(first, second, options.update === true, options.force === true,options.own === true, options.start);
    })
    

program.parse();

commands/checkVersion.js
const greaterVersion = (v_old, v_new) => {
    let res = {
        major : false,
        minor : false,
        patch : false
    };
    if(v_old[0] > v_new[0]){        //existing version is major update(s) ahead
        return res
    }
    if(v_old[0] < v_new[0]){        //existing version is major update(s) behind
        return {...res, major : true}
    }
    if(v_old[1] > v_new[1]){        //existing version is minor update(s) ahead
        return res
    }
    if(v_old[1] < v_new[1]){        //existing version is minor update(s) behind
        return {...res, minor : true}
    }
    if(v_old[2] >= v_new[2]){       //existing version is patch update(s) ahead or is at same version
        return res
    }
    return {...res,patch : true}    //existing version is patch update(s) behind
}

const checkVersion = (version_old, version_new) => {
    version_new = version_new.split('.').map(Number)
    let versionInfo = new Array(new Array());
    if(version_old.includes('-')){
        versionInfo[0] = version_old.split('-')[0].split('.').map(Number);
        versionInfo[1] = version_old.split('-')[1].split('.').map(Number);
        let lowerResponse = greaterVersion(versionInfo[0], version_new);
        let lowerFlag = false;
        for (const key in lowerResponse) {
            if(lowerResponse[key]){
                lowerFlag = true;
            }
        }
        if(!lowerFlag){     //below lower range -> requires no update
            return {
                satisfied : true
            }
        }
        let upperResponse = greaterVersion(version_new, versionInfo[1]);
        let upperFlag = false;
        for (const key in upperResponse) {
            if(upperResponse[key]){
                upperFlag = true;
            }
        }
        if(!upperFlag){     //below upper range -> requires normal update
            return {
                satisfied : false
            }
        }else{              //above upper range -> requires force update
            return {
                satisfied : false,
                forceRequired : true
            }
        }
    }else if(version_old.charAt() === '^' || version_old.charAt() === '~'){
        versionInfo[0] = version_old.slice(1).split('.').map(Number);
        let gtResponse = greaterVersion(versionInfo[0], version_new);
        if(!(gtResponse["major"] || gtResponse["minor"] || gtResponse["patch"])){          //above specified version -> requires no update
            return {
                satisfied : true
            }
        }
        if(gtResponse["major"]){      //requires major update
            return {
                satisfied : false,
                forceRequired : true
            }
        }else if(version_old.charAt() === '~' && gtResponse["minor"]){    //requires minor update
            return {
                satisfied : false,
                forceRequired : true
            }
        }
        else{                          //requires normal update
            return {
                satisfied : false
            }
        }
    }
    else if((version_old.split('.').length - 1) == 0){
        versionInfo[0] = version.split('.').map(Number).concat([0,0]);
        let gtResponse = greaterVersion(versionInfo[0], version_new);
        if(!(gtResponse["major"] || gtResponse["minor"] || gtResponse["patch"])){          //above specified version -> requires no update
            return {
                satisfied : true
            }
        }
        if(gtResponse["major"]){      //requires major update
            return {
                satisfied : false,
                forceRequired : true
            }
        }else{
            return {            //requires normal update
                satisfied : false
            }
        }
    }else if((version_old.split('.').length - 1) == 1){
        versionInfo[0] = version.split('.').map(Number).concat([0]);
        let gtResponse = greaterVersion(versionInfo[0],version_new);
        if(!(gtResponse["major"] || gtResponse["minor"] || gtResponse["patch"])){          //above specified version -> requires no update
            return {
                satisfied : true
            }
        }
        if(gtResponse["major"] || gtResponse["minor"]){      //requires major or minor update
            return {
                satisfied : false,
                forceRequired : true
            }
        }
        else{                          //requires normal update
            return {
                satisfied : false
            }
        }
    }
}

export default checkVersion;

commands/updateDependency.js
import { Octokit } from "@octokit/rest";
import dotenv from 'dotenv';
import {execSync} from 'child_process'
import path from "path";
import fs from "fs";
import chalk from "chalk";

dotenv.config();

const octokit = new Octokit({
    auth: process.env.G_TOKEN
})

const fork = async(owner,repo) => {
  console.log(chalk.cyan('\nTrying to fork the repo\n'))
  try{
    const response = await octokit.request(`POST /repos/${owner}/${repo}/forks`, {
    owner: owner,
    repo: repo
    })
    if(response.status === 202){
      console.log(chalk.green('\nSuccessfully forked the repo\n'))
    }else{
      throw new Error('received wrong status code',{cause : response})
    }
    return response;
  }catch(err){
    throw new Error('Something went wrong during trying to fork the repo.', {cause : err});
  }
}

const clone_and_push = (repository, pckg, repo) => {
  try{
    fs.mkdirSync('./Temporary_Directory')
    console.log(chalk.green('Created a temporary directory with the name "Temporary Directory". This is where the repo will be cloned.'))
  }catch(err){
    if(err.code !== 'EEXIST'){
      console.log(err)
      throw new Error('There seems to be a problem while working with the filesystem.', {cause : err})
    }else if(err.code === 'EEXIST'){
      console.log(chalk.yellow(`\nThe repo will be cloned into the "Temporary Directory\\${repo}"\n`))
    }
  }
  try{
    console.log(chalk.cyan("\nTrying to clone the repo\n"))
    process.chdir('./Temporary_Directory')
    execSync(`git clone --filter=tree:0 --sparse --depth=1 ${repository + '.git'}`, {
      stdio : [0,1,2],
      cwd : path.resolve(process.cwd(), '')
    })
    console.log(chalk.cyan('\nCloning Done. Now Creating a new branch with the name "updates"\n'))
    process.chdir(`./${repo}`);
    execSync('git checkout -b updates', {
      stdio : [0,1,2],
      cwd : path.resolve(process.cwd(), '')
    })
    console.log(chalk.green("\nSuccessfully created a new branch\n"))
    console.log(chalk.cyan(`\nNow Updating the package to ${pckg}\n`))
    execSync(`npm i ${pckg}`, {
      stdio : [0,1,2],
      cwd : path.resolve(process.cwd(), '')
    })
    console.log(chalk.cyan(`\nUpdated the package to ${pckg}\n`))
    console.log(chalk.cyan('\nNow trying to commit...\n'))
    execSync('git add .', {
      stdio : [0,1,2],
      cwd : path.resolve(process.cwd(), '')
    })
    execSync(`git commit -m "auto updated to ${pckg}"`, {
      stdio : [0,1,2],
      cwd : path.resolve(process.cwd(), '')
    })
    console.log(chalk.cyan('\nNow trying to push to origin...\n'))
    execSync("git push origin updates", {
      stdio : [0,1,2],
      cwd : path.resolve(process.cwd(), '')
    })
    console.log(chalk.green("\nSuccessfully cloned the repo, created a branch, updated the package and pushed to origin.\n"))

    process.chdir('../../')
    return true;
  }catch(err){
    throw new Error('Something went wrong during trying to clone undate and push.', {cause : err})
  }
  
}

const pull = async (owner, repo,login, pckg) => {
  console.log(chalk.cyan('\nTrying to generate a PR(Pull Request)\n'))
  try{
    let {data : {default_branch}} = await octokit.request(`GET /repos/${owner}/${repo}`, {
      owner: owner,
      repo: repo
    })
    const response = await octokit.request(`POST /repos/${owner}/${repo}/pulls`, {
      owner: owner,
      repo: repo,
      title: `Upgraded a package named ${pckg}`,
      body: 'Please pull these awesome changes in!',
      head: `${login}:updates`,
      base: default_branch
    })
    if(response.status === 201){
      console.log(chalk.green(`\nSuccessfully generated a PR at ${owner}/${repo}...Hopefully the author will see it and value your contribution. ;)\n`))
      return response
    }else{
      throw new Error('Received Wrong Status Code', {cause : response})
    }
  }catch(err){
    throw new Error('Something went wrong during trying to create a Pull Request.', {cause : err})
  }
  
}

const updateDependency = async (user, repo, pckg, ownRepo) => {
    try{
      let forkResponse;
      if(!ownRepo){
        forkResponse = await fork(user, repo);
      }
      const repoLinkToClone = forkResponse?.data["html_url"] || `https://github.com/${user}/${repo}`;
      let cloneResponse = clone_and_push(repoLinkToClone, pckg, repo);
      let {data : {login}} = await octokit.request('GET /user', {})
      let pullResponse = await pull(user, repo, login, pckg)
      return pullResponse.data["html_url"];
    }catch(err){
      console.log(chalk.red(`\nSomething looks suspicious. Here take a look at the error ${err}\n`))
    }
}


export {
    fork,
    clone_and_push,
    pull,
    updateDependency
};